corgi/tasks/managed_services.py

```diff
@@ -59,42 +79,60 @@ def cpu_manifest_service(product_stream_id: str, service_components: list) -> No
             continue
 
         with transaction.atomic():
-            # Find previous build and delete it and its components, so we can create a fresh
-            # manifest structure for this "new" build. This is done specifically because we don't
-            # have a way to tie a set of components to a specific build right now,
-            # so we construct an arbitrary build periodically even if nothing has changed since
-            # the last time we looked. Historical data is not needed as of right now.
-            past_builds = SoftwareBuild.objects.filter(
-                name=service_component["name"],
-                build_type=SoftwareBuild.Type.APP_INTERFACE,
-                meta_attr__service=service.name,
-            )
-            for build in past_builds:
-                ProductComponentRelation.objects.filter(software_build=build).delete()
-                build.delete()
-
             build_id = now.strftime("%Y%m%d%H%M%S%f")
-            build = SoftwareBuild.objects.create(
-                name=service_component["name"],
-                build_type=SoftwareBuild.Type.APP_INTERFACE,
-                build_id=build_id,
-                completion_time=now,
-                meta_attr={"service": service.name},
-            )
-
-            root_component = Component.objects.create(
+            try:
+                # get_or_create requires all keyword arguments to appear in
+                # a unique_together constraint, like build_id and build_type
+                # Otherwise it will not be atomic and can behave incorrectly
+                # Since we don't have a database index for name + build_type,
+                # we try to find an APP_INTERFACE build with a matching name first
+                # And update the list of services it belongs to
+                # If that fails, we'll create a new build
+                # So Corgi reuses a component when services reuse the component
+                build = SoftwareBuild.objects.get(
```

Comment:
In what scenario will this actually return a build when all builds are deleted before these tasks run? Are there really services that share the same component (Quay image / GitHub repo)?